Blog post part 1 about MyST and Nikola

damianavila opened the pull request:
This is the first part of a blog series about my deep dive into the MyST world.
You may notice the colloquial way this is written because this was intended to be published in my blog.
Related to this, this is a md version living in a PR so others can see it and review it.
If this one ends up in my blog, I will convert it to a ipynb file before publication.

damianavila committed:
Add files via upload

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+# A deep dive into the MyST world
```
```suggestion
# A deep dive into MyST Markdown, Part 1: Roles and directives in Nikola
```

choldgraf replied:
I think this would make the title a bit more specific so that you could build upon it for your other parts. (feel free to re-work the wording if you like)

damianavila replied:
> (feel free to re-work the wording if you like)

OK, let me think about this one...

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+# A deep dive into the MyST world
+
+> *This is a series of blog posts inviting you to join me in a little journey I have experienced in the last few weeks at the time to figure it out a nice story for **MyST**
```
Just a meta thought - I like using the markdown approach of _linebreak after every sentence_ so that you don't run into semi-arbitrary line breaks like this

damianavila replied:
Mmm... I need to try it... I see how linebreak after every sentence could be nice for some people and really ugly for others. I promise I will eventually make a try with another piece of text :wink:

choldgraf replied:
in case it's helpful, I've found this site to be a useful rationale for doing line breaks this way :-) https://sembr.org/

they actually recommend more complex rules for deciding when to break lines, but I think that "one line per sentence" is easy enough to follow

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+Hey, a lot of time without writing, you know... (lack of) time, but let's go back to the
+important things... or at least, interesting stuff to share with you!
+
+Recently, I have been involved with the [EBP](https://executablebooks.org/en/latest/)
```
```suggestion
Recently, I have been involved with the [Executable Books](https://executablebooks.org/en/latest/)
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+Recently, I have been involved with the [EBP](https://executablebooks.org/en/latest/)
+project and community (you can learn more about them [here](https://executablebooks.org/en/latest/about.html)). One of the major projects inside that community is the
+[MyST-Parser](https://github.com/executablebooks/MyST-Parser), a *rich* and *extensible*
+flavor of Markdown meant for technical documentation and publishing. It supports all thesyntax of CommonMark Markdown at its base but, in addition, it includes several
```
```suggestion
flavor of Markdown meant for technical documentation and publishing. It supports all the syntax of CommonMark Markdown at its base but, in addition, it includes several
```

choldgraf replied:
there are a few other places where you've forgotten a space, I'll try to find them but you should double-check

damianavila replied:
OK, I will pay attention to other occurrences!

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+extensions to CommonMark adding extra syntax features for technical writing, such as the
+roles and directives used by Sphinx in the rST world!
+
+When I looked into this project, I quickly realized it would interesting to explore how
```
I'd try and reduce the number of adverbs that you use. I know that you are shooting for a "familiar" tone here, and in general they are OK, but using them so much makes it a bit harder to follow the flow of sentences.

```suggestion
When I looked into this project, I realized it would interesting to explore how
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+this flavor of Markdown could potentially integrate with the [Nikola](https://getnikola.com/)
+ecosystem. If you look into my previous blog posts, you will see I am a Nikola
+user and contributor and I am always interested in how we can improve the
+"blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circle
```
A meta-question: did you also look into using Jupytext for blogging with MyST Jupyter Notebooks in Nikola? This could be a nice way to keep the structure of notebooks, but with the benefit of version control

damianavila replied:
I briefly look into it but did not go full on that because it was not giving enough exposure to the MyST machinery, IMHO.
I think it would be a really nice topic for another (and related) blog post, actually.

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+this flavor of Markdown could potentially integrate with the [Nikola](https://getnikola.com/)
+ecosystem. If you look into my previous blog posts, you will see I am a Nikola
+user and contributor and I am always interested in how we can improve the
+"blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circle
+around an excel pipeline for publishing Jupyter Notebook being the Myst-Parser one of
+the central building blocks. They are even [interested](https://github.com/executablebooks/jupyter-book/issues/900)
```
```suggestion
"blogging story" for those using Jupyter Notebooks. And the whole **EBP** project circles
around an excellent pipeline for publishing Jupyter Notebook, with the MyST-Parser being one of
the central building blocks. They are even [interested](https://github.com/executablebooks/jupyter-book/issues/900)
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+
+# First impressions (someone has beaten me to it...)
+
+Ok, the first idea coming to my mind was to somehow implement MyST support in Nikola.
```
Are you OK with me making small english changes like this? 
```suggestion
The first idea that came to my mind was to somehow implement MyST support in Nikola.
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+Ok, the first idea coming to my mind was to somehow implement MyST support in Nikola.
+But, hey, somebody probably already thought about this before me. So I made a search to
+see if this was not previously explored by others. And indeed that was the case!
+Roberto, the Nikola creator, wrote a plugin for MyST about a year ago (there is a brief
+blog post talking about it [here](https://ralsina.me/tr/es/weblog/posts/new-plugins-for-nikola.html)).
```
Try to hyperlink more of the text so that it is easier for people to notice. For example, [the sentence here](https://executablebooks.org) is easier to notice, whereas this sentence [here](https://executablebooks.org) is harder to notice.
```suggestion
Roberto, the Nikola creator, wrote a plugin for MyST about a year ago (there is [a brief
blog post talking about it](https://ralsina.me/tr/es/weblog/posts/new-plugins-for-nikola.html)).
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+Problem solved! Someone else had the idea, implemented it and now I can use it, right?
+Well... actually it was not that simple...
+
+# Current limitations (there is always a but...)
```
This should be a sub-section, right? 
```suggestion
## Current limitations (there is always a but...)
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+
+> Heya, yes it is; you have to use sphinx to fully parse to HTML
+
+Wait? Why? I was expecting the MyST Python API to produce the expected HTML for roles
+and directives and that is not supported? Is the API incomplete? Is this intended by
+design? A lot of questions came to my mind and, as usual, the answer was in the
+documentation and the codebase.
```
Just a note that this might come across as kind-of negative on the myst parser...it might be better to just get straight to the point, e.g. "This surprised me, as I assumed that roles and directives would be supported by the Python API, but I learned that it is more complicated than that."

damianavila replied:
Thank you... the negativity you mentioned did not pass through my head at any time. I will do the change as requested.

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+Those are an integral part of the Myst-Parser project. But when you look at the `html`
+renderer, that [RendererHTML] class (https://markdown-it-py.readthedocs.io/en/latest/api/markdown_it.renderer.html#markdown_it.renderer.RendererHTML)
+is coming from another EBP project called markdown-it-py. This is a Python port of the
+well-used [markdown-it](https://github.com/markdown-it/markdown-it), plus some of itsassociated plugins.
```
```suggestion
well-used [markdown-it](https://github.com/markdown-it/markdown-it), plus some of its associated plugins.
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+        .use(footnote_plugin)
+        .use(wordcount_plugin, per_minute=config.words_per_minute)
+        .disable("footnote_inline")
+        # disable this for now, because it need a new implementation in the renderer
```
Just to avoid readers needing to wonder what this comment means
```suggestion
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,174 @@
+# Overcome the limitations (don't worry I will stop here... for now)
+
+OK, this is long enough for now, the next post will deep dive a little bit more in the
+Myst-parser and the markdown-it-py codebases in the context of potentially supporting
```
```suggestion
MyST-parser and the markdown-it-py codebases in the context of potentially supporting
```

choldgraf replied:
I think it should be either `myst` (all lowercase) or `MyST`, but not intermixed (`Myst`)

damianavila replied:
I totally agree!

choldgraf reviewed (commented):
Thanks for sharing this post @damianavila - I've left a few comments and minor suggested edits throughout. In general it does a nice job of explaining your perspective on the parser as you played around with it and learned more.

The one major suggestion I'd give is to find ways to reduce your word-count. There are a few sections where you've got a lot of "fluff" before sentences, which is fine when used sparingly, but can make it harder to follow sentences when there is a lot of it. Even though this is for a personal blog, I find that it is better to get straight to the point relatively quickly, so that your reader doesn't have to scan a lot to find what they're interested in.

happy to take another pass if you'd like, but feel free to publish

damianavila committed:
Use Executable Books name instead of an acronym

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Add missing space

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Remove "quickly" from the sentence

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Use excellent instead... and some other styles fixes.

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Fix style

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Fix linking.

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Add another level

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Add missing space

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Remove commented line

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Use the proper acronym

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Remove fluff and try to reduce word count

The main idea of this commit was to reduce the word count removing the
"fluff" before some sentences. I have also reordered some paragraphs and
remove any potentially kind -pf negativity about the MyST Python API.

Thanks choldgraf for the review and advice!

damianavila commented:
OK, this was my first editorial review in one of my blog post!
Thanks @choldgraf!!

Btw, can you make another pass on it?

Also, do you have any problems if I thank you at the end of the post as well for the review? :wink:

choldgraf commented:
Sounds good! Can you put it in the review column so i know to take a look? 🙂

And I'm happy to help! I was an editor at the Berkeley science review for 6 years so I hope this is useful feedback!

damianavila commented:
> Sounds good! Can you put it in the review column so i know to take a look? 🙂

Done!

>And I'm happy to help! I was an editor at the Berkeley science review for 6 years so I hope this is useful feedback!

Super useful! Thanks again!

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,167 @@
+# A deep dive into MyST, Part 1: The MyST-Parser Python API usage in Nikola
+
+> *This is a series of blog posts inviting you to join me in a little journey I have
+experienced in the last few weeks at the time to figure it out a nice story for **MyST**
+inside the **Nikola** world.*
```
```suggestion
> *This is a series of blog posts inviting you to join me in a little journey I have
experienced in the last few weeks to figure out a nice story for **MyST**
inside the **Nikola** world.*
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,167 @@
+technical writing, such as the roles and directives used by Sphinx in the rST world!
+
+The whole Executable Books project is about providing an excellent pipeline for
+publishing Jupyter Notebook, with the MyST-Parser being one of the central building
```
```suggestion
publishing Jupyter Notebooks, with the MyST-Parser being one of the central building
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,167 @@
+
+I started digging into the codebase and some MyST documentation.
+
+If you look at the `to_html` function I shown before: 
```
```suggestion
If you look at the `to_html` function I showed before: 
```

choldgraf commented on posts/draft.md:
```diff
@@ -0,0 +1,167 @@
+In this model, it actually makes sense to see roles and directives supported when you go
+through the Docutils/Sphinx path (those concepts are core-supported in the
+Docutils/Sphinx ecosystem) and those unsupported when you use the MyST Python API which
+does not understand how to parse nor render these concepts as we previously shown.
```
```suggestion
does not understand how to parse nor render these concepts as we've previously shown.
```

choldgraf reviewed (approved):
This reads much more cleanly to me! Just a few small comments in there, publish when you like!

damianavila committed:
Remove unnecessary words

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Add plural

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Correct past tense

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila committed:
Add missing verb contraction

Co-authored-by: Chris Holdgraf <choldgraf@gmail.com>

damianavila commented:
> This reads much more cleanly to me!

Super!

> Just a few small comments in there, 

Accepted!

> publish when you like!

Publishing on Monday afternoon (I do not like to publish on Fri afternoon when everyone is closing their computers :wink:),

choldgraf commented:
wohooo!
